src/cpp/session/modules/environment/SessionEnvironment.cpp

```diff
@@ -119,95 +120,192 @@ bool handleRBrowseEnv(const core::FilePath& filePath)
-   if (visited.count(obj))
+   if (obj == R_NilValue || visited.count(obj))
       return false;
 
    // mark SEXP as visited
    visited.insert(obj);
 
-   // list the contents of this environment
-   std::vector<r::sexp::Variable> vars;
-   r::sexp::Protect rProtect;
-   if (TYPEOF(obj) == S4SXP)
+   // check if this is an external pointer
+   if (r::sexp::isExternalPointer(obj)) 
    {
-      // for S4 objects, list the attributes (which correspond to slots)
-      r::sexp::listNamedAttributes(obj, &rProtect, &vars);
+      // NOTE: this includes UserDefinedDatabase, aka 
+      //       external pointers to R_ObjectTable
+
+      // when nullPtr is true, only return true for null pointer xp
+      // otherwise only return true for non null pointer xp
+      if (nullPtr == (r::sexp::getExternalPtrAddr(obj) == nullptr))
+         return true;
+
+      if (hasExternalPointer(EXTPTR_PROT(obj), nullPtr, visited))
+         return true;
+
+      if (hasExternalPointer(EXTPTR_TAG(obj), nullPtr, visited))
+         return true;
    }
-   else
+
+   switch(TYPEOF(obj))
    {
-      // not S4, coerce to environment
-      SEXP envir = R_NilValue;
-      if (TYPEOF(obj) == ENVSXP)
+      case ENVSXP: 
       {
-         // we were given a primitive environment (ENVSXP)
-         envir = obj;
+         if (envHasExternalPointer(obj, nullPtr, visited))
+            return true;
+         break;
       }
-      else
+      case VECSXP:
+      case EXPRSXP:
       {
-         // convert the passed environment into a primitive environment; this is required so that
-         // e.g. reference objects that subclass 'environment' can be introspected below
-         Error error = r::sexp::asPrimitiveEnvironment(obj, &envir, &rProtect);
-         if (error)
+         if (listHasExternalPointer(obj, nullPtr, visited))
+            return true;
+         break;
+      }
+         
+      case LISTSXP:
+      case LANGSXP:
+      {
+         if (pairlistHasExternalPointer(obj, nullPtr, visited))
+            return true;
+         break;
+      }
+         
+      case WEAKREFSXP:
+      {
+         if (weakrefHasExternalPointer(obj, nullPtr, visited))
+            return true;
+
+         break;
+      }
+      case PROMSXP: 
+      {
+         SEXP value = PRVALUE(obj);
+         if (value != R_UnboundValue)
+         {
+            if (hasExternalPointer(value, nullPtr, visited))
+               return true;
+         }
+         else 
          {
-            // can't search in here
+            if (hasExternalPointer(PRCODE(obj), nullPtr, visited))
+               return true;
+
+            if (hasExternalPointer(PRENV(obj), nullPtr, visited))
+               return true;
+
             return false;
          }
+         break;
       }
-
-      r::sexp::listEnvironment(envir,
-                               true,  // include all values
-                               false, // don't include last dot
-                               &vars);
+      default:
```

Comment:
should there be something for `CLOSXP` presumabmly checking `FORMALS`, `BODY` and `CLOENV` (its parents ?) for external pointers ?

Reply:
but maybe that's overkill